pkg/virt-handler/retry_manager_test.go

```diff
@@ -0,0 +1,97 @@
+)
+
+var _ = Describe("virt-handler retry manager", Ordered, func() {
+	initialWait, maxWait, maxFailResponseTime := 200*time.Millisecond, 5*time.Second, 1*time.Second
+	buffer := 10 * time.Millisecond
+	var retryManager *FailRetryManager
+	key := "c4ab4ae0-db63-45d8-aa0f-fc53dc84bdab"
+
+	Context("retry manager", func() {
+
+		BeforeEach(func() {
+			retryManager = NewFailRetryManager("test", initialWait, maxWait, maxFailResponseTime)
+		})
+
+		It("Should not delay if no failure occurred", func() {
+			shouldDelay, _ := retryManager.ShouldDelay(key, func() bool { return false })
+			Expect(shouldDelay).To(BeFalse())
+		})
+
+		It("Should not delay on first failure", func() {
+			shouldDelay, _ := retryManager.ShouldDelay(key, func() bool { return true })
+			Expect(shouldDelay).To(BeFalse())
+			Expect(retryManager.retryStates[key].lastRunFailed).To(BeFalse())
+
+			// Call again as a failure response to the above call.
+			shouldDelay, _ = retryManager.ShouldDelay(key, func() bool { return true })
+			Expect(shouldDelay).To(BeTrue())
+			Expect(retryManager.retryStates[key].lastRunFailed).To(BeTrue())
+		})
+
+		It("Should do exponential backoff on failure retries", func() {
+			shouldDelay, _ := retryManager.ShouldDelay(key, func() bool { return true })
+			Expect(shouldDelay).To(BeFalse())
+			Expect(retryManager.retryStates[key].lastRunFailed).To(BeFalse())
+			// Call again as a failure response to the above call.
+			shouldDelay, _ = retryManager.ShouldDelay(key, func() bool { return true })
+			Expect(shouldDelay).To(BeTrue())
+			Expect(retryManager.retryStates[key].lastRunFailed).To(BeTrue())
+
+			//check waitInterval before the timer ends
+			time.Sleep(initialWait - buffer)
+			shouldDelay, _ = retryManager.ShouldDelay(key, func() bool { return true })
+			Expect(shouldDelay).To(BeTrue())
+			Expect(retryManager.retryStates[key].waitInterval).To(BeEquivalentTo(initialWait))
+
+			//check waitInterval just after the timer ends
+			time.Sleep(buffer)
+			shouldDelay, _ = retryManager.ShouldDelay(key, func() bool { return true })
+			Expect(shouldDelay).To(BeFalse())
+			Expect(retryManager.retryStates[key].waitInterval).To(BeEquivalentTo(initialWait * 2))
+		})
+
+		It("Should reset when no failure is reported within the max response time", func() {
+			shouldDelay, _ := retryManager.ShouldDelay(key, func() bool { return true })
+			Expect(shouldDelay).To(BeFalse())
+			_, firstEventFirstFail := retryManager.GetFirstFailureTime(key)
+
+			time.Sleep(maxFailResponseTime + buffer)
+			shouldDelay, _ = retryManager.ShouldDelay(key, func() bool { return true })
+			Expect(shouldDelay).To(BeFalse())
+			_, lastEventFirstFail := retryManager.GetFirstFailureTime(key)
+			Expect(lastEventFirstFail).ToNot(BeEquivalentTo(firstEventFirstFail))
+		})
+
+		It("Should be able to remove a key", func() {
+			retryManager.Remove(key)
+			found, _ := retryManager.GetFirstFailureTime(key)
+			Expect(found).To(BeFalse())
+		})
```

Comment:
I see that the `Remove()` method is used only here. Why do we need it if the keys are being garbaged-collected?

Reply:
Thanks! Initially I think I need it, but at the end it is unused. I removed it, and `GetFirstFailureTime` too (used only in test too)